supertokens_python/recipe/multifactorauth/utils.py

```diff
@@ -0,0 +1,221 @@
+                        computed_first_factor_id_for_session = FactorIds.EMAILPASSWORD
+                        break
+                elif login_method.recipe_id == "thirdparty":
+                    valid_res = await is_valid_first_factor(
+                        tenant_id, FactorIds.THIRDPARTY, user_context
+                    )
+                    if valid_res.status == "TENANT_NOT_FOUND_ERROR":
+                        raise SuperTokensSessionError(
+                            SuperTokensSessionError.UNAUTHORISED, "Tenant not found"
+                        )
+                    elif valid_res.status == "OK":
+                        computed_first_factor_id_for_session = FactorIds.THIRDPARTY
+                        break
+                else:
+                    factors_to_check = []
+                    if login_method.email is not None:
+                        factors_to_check.extend(
+                            [FactorIds.LINK_EMAIL, FactorIds.OTP_EMAIL]
+                        )
+                    if login_method.phone_number is not None:
+                        factors_to_check.extend(
+                            [FactorIds.LINK_PHONE, FactorIds.OTP_PHONE]
+                        )
+
+                    for factor_id in factors_to_check:
+                        valid_res = await is_valid_first_factor(
+                            tenant_id, factor_id, user_context
+                        )
+                        if valid_res.status == "TENANT_NOT_FOUND_ERROR":
+                            raise SuperTokensSessionError(
+                                SuperTokensSessionError.UNAUTHORISED, "Tenant not found"
+                            )
+                        elif valid_res.status == "OK":
+                            computed_first_factor_id_for_session = factor_id
+                            break
+
+                    if computed_first_factor_id_for_session is not None:
+                        break
+
+        if computed_first_factor_id_for_session is None:
+            raise SuperTokensSessionError(
+                SuperTokensSessionError.UNAUTHORISED, "Incorrect login method used"
+            )
+
+        updated_claim_val = True
+        mfa_claim_value = {
+            "c": {computed_first_factor_id_for_session: first_factor_time},
+            "v": True,  # updated later in this function
+        }
+
+    completed_factors = mfa_claim_value["c"]
+
+    user_prom: Optional[asyncio.Future] = None
+
+    async def user_getter():
+        nonlocal user_prom
+        if user_prom is None:
+            user_prom = asyncio.Future()
+            session_user = await get_user(session_recipe_user_id, user_context)
+            if session_user is None:
+                raise SuperTokensSessionError(
+                    SuperTokensSessionError.UNAUTHORISED, "Session user not found"
+                )
+            user_prom.set_result(session_user)
+        return await user_prom
```

Comment:
This is incorrect translation. We simply store the user object instead of user_prom (initially its set to None). Then when this function is called, we check if the user object is None, and if it is, we call the get_user function and then save the resulting user object in the variable, and return that. No need for asyncio use.

Reply:
Implemented using task